docs/decisions/0023-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,267 @@
+    - Dimensions: ComponentType, ComponentName, Service ID, Model ID
+    - Example:
+      | ComponentType | ComponentName | Fail | Success
+      |---|---|---|---|
+      | Plan | HandlebarsPlan | 5 | 95
+      | Plan | SequentialPlan | 20 | 80
+
+## Considered Options
+
+- Function hooks
+  - Inject logic to functions that will get executed before or after a function is invoked.
+- Instrumentation
+  - Logging
+  - Metrics
+  - Traces
+
+## Other Considerations
+
+SK currently tracks token usage metrics in connectors; however, these metrics are not categorized. Consequently, developers cannot determine token usage for different operations. To address this issue, we propose the following two approaches:
+
+- Bottom-up: Propagate token usage information from connectors back to the functions, along with model results.
+- Top-down: Propagate function information down to the connectors, enabling them to tag metric items with function information.
+
+We have decided to implement the bottom-up approach for the following reasons:
+
+1. SK is already configured to propagate token usage information from connectors via `ModelResult`. We simply need to extend the list of items that get propagated, such as model information.
+2. Currently, SK does not have a method for passing function information down to the connector level. Although we considered using [baggage](https://opentelemetry.io/docs/concepts/signals/baggage/#:~:text=In%20OpenTelemetry%2C%20Baggage%20is%20contextual%20information%20that%E2%80%99s%20passed,available%20to%20any%20span%20created%20within%20that%20trace.) as a means of propagating information downward, experts from the OpenTelemetry team advised against this approach due to security concerns.
+
+With the bottom-up approach, we need to add model information to `IResultBase`:
+
+```csharp
+/// <summary>
+/// Interface for model results
+/// </summary>
+public interface IResultBase
+{
+  /// <summary>
+  /// Model name or Id
+  /// </summary>
+  string Model { get; }
+
+  /// <summary>
+  /// Gets the model result data.
+  /// </summary>
+  ModelResult ModelResult { get; }
+}
+```
+
+## Decision Outcome
+
+TBD
+
+## Validation
+
+Tests can be added to make sure that all the expected telemetry items are in place and of the correct format.
+
+## Description the Options
+
+### Function hooks
+
+Function hooks allow developers to inject logic to the kernel that will be executed before or after a function is invoked. Example use cases include logging the function input before a function is invoked, and logging results after the function returns.
+For more information, please refer to the following ADRs:
+
+1. [Kernel Hooks Phase 1](./0005-kernel-hooks-phase1.md)
+2. [Kernel Hooks Phase 2](./0018-kernel-hooks-phase2.md)
+
+We can inject, during function registration, default callbacks to log critical information for all functions.
```

Comment:
All functions now derivs from KernelFunction, and we can put arbitrary logic into the base implementation. That's where metrics are now tracked for function duration, errors, etc.

Reply:
This makes it even better! Thanks!